src/function/gds/output_writer.cpp

```diff
@@ -104,99 +107,139 @@ void PathsOutputWriter::write(processor::FactorizedTable& fTable, nodeID_t dstNo
                 if (curPath.size() == 1) {
                     setLength(lengthVector.get(), curPath[0]->getIter());
                 }
-            } else {
-                curPath.pop_back();
+                backtracking = false;
+                break;
             }
         } else {
-            auto parent = bfsGraph.getInitialParentAndNextPtr(topNodeID);
-            while (parent->getIter() != top->getIter() - 1) {
+            auto top = getTop(curPath);
+            auto parent = bfsGraph.getInitialParentAndNextPtr(top->getNodeID());
+            while (true) {
+                if (parent == nullptr) {
+                    // No more forward tracking candidates. Switch to backward tracking.
+                    backtracking = true;
+                    break;
+                }
+                if (parent->getIter() == top->getIter() - 1 &&
+                    checkNewPathElement(curPath, parent)) {
+                    // A forward tracking candidate should decrease the iteration by one and also
+                    // pass node predicate checking.
+                    curPath.push_back(parent);
+                    backtracking = false;
+                    break;
+                }
                 parent = parent->getNextPtr();
             }
-            curPath.push_back(parent);
-            backtracking = false;
         }
     }
 }
 
-bool PathsOutputWriter::checkPathNodeMask(const std::vector<ParentList*>& path) const {
-    if (!info.hasNodeMask()) {
+ParentList* PathsOutputWriter::findFirstParent(offset_t dstOffset, BFSGraph& bfsGraph) const {
+    auto result = bfsGraph.getCurFixedParentPtrs()[dstOffset].load(std::memory_order_relaxed);
+    while (result) {
+        // A valid parent should
+        // (1) satisfies path node semi mask (i.e. path node predicate)
+        // (2) since first parent has the largest iteration number which decides path length, we
+        //     also need to check if path length is greater than lower bound.
+        if (checkPathNodeMask(result) && result->getIter() >= info.lowerBound) {
+            break;
+        }
+        result = result->getNextPtr();
+    }
+    return result;
+}
+
+// This code checks if we should switch from backtracking to forward-tracking, i.e.,
+// moving forward in the DFS logic to find paths. We switch from backtracking if:
+bool PathsOutputWriter::canNextReplacePathTop(ParentList* next,
+    const std::vector<ParentList*>& path) const {
+    if (next == nullptr) {
+        return false;
+    }
+    auto nextIter = next->getIter();
+    // (1) if this is the first element in the stack (curPath.size() == 1), i.e., we
+    // are enumerating the parents of the destination, then we should switch to
+    // forward-tracking if the next parent has visited the destination at a length
+    // that's greater than or equal to the lower bound of the recursive join. Otherwise,
+    // we'll enumerate paths that are smaller than the lower; OR
```

Comment:
Let's be more explicit on this comment (which I know I wrote before): "... Otherwise, we would enumerate paths that are smaller than the lower bound from the start element, so we can stop here.

I also think there is a bug here. The code should be:
```
    if (path.size() == 1) {
        return nextIter >= info.lowerBound;
    }
```
If the path.size() == 1, then you should return true if nextIter >= info.lowerBound and return false otherwise. You are not returning false. This may not strictly be a bug but let's be explicit in the logic of the enumeration.

Reply:
Done